### ISSUE TITLE
Rule proposal: mistaken parentheses

### ISSUE THREAD

**Issue by rightaway** [sindresorhus/eslint-plugin-unicorn#1796]
### Description

There's a certain type of errors that can be hard to spot and will work fine. Since it's valid to do operations like `||` in this context is it even possible to have a rule that could notice when you might have meant the other one?

### Fail

This is what you wrote

```js
str.startsWith('str' || str.startsWith('str'))
```

### Pass

This is what you meant

```js
str.startsWith('str') || str.startsWith('str')
```

**Comment by sindresorhus** [comment]
@rightaway Can you suggest a name for the rule?

**Comment by rightaway** [comment]
`mistaken-parentheses-nesting`?

Which of the following cases would be reported as errors by this rule?

```js
func(arg || func(arg))
func(arg || differentfunc(arg))
func(arg || func(differentarg))
func(arg || differentfunc(differentarg))
```

**Comment by fisker** [comment]
`no-suspicious-parentheses`?

**Comment by sindresorhus** [comment]
`no-confusing-parens`?

ESLint uses the word "confusing" in a few rules.

**Comment by fregante** [comment]
I don't see a problem with the code `x(y || z())` and it can't be reported as a bug. The only issue that can be detected is when conditions are constant, which actually already exists as a core rule (but likely doesn't catch this case)

https://eslint.org/docs/rules/no-constant-condition

The failing part would be `'str' || anythingElse` since `anythingElse` is unreachable.

**Comment by fisker** [comment]
I think we should only check built-in functions that doesn't accept a `boolean`.

**Comment by sindresorhus** [comment]
Accepted. Name should be `no-confusing-parens`.

**Comment by sindresorhus** [comment]
The motivating example is covered by ESLint core `no-constant-binary-expression`, which catches logical expressions that always short-circuit.

The broader form (`func(arg || func(arg))`) is too ambiguous without type information. A Unicorn rule would either need a large hand-maintained table of built-in argument types or would report valid fallback expressions. That does not feel worth adding as `no-confusing-parens`.
